### PR TITLE
Add Basset support

### PR DESCRIPTION
**Pull request by abompard**
This PR implements the Noggin-side of the workflow described in #27 .

- Users are created with the `spamcheck_awaiting` status.
- Upon registration, a request is sent to Basset to check for spam likelihood. In addition to the user details, this request contains a callback URL and a token that Basset will have to provide with the result
- A webhook endpoint is added to receive Basset's result, which will update the user's status.
- When a user status is not `active`, a landing page is provided to explain to the user that their account is being spamchecked.

### CODE REVIEW

File: noggin/representation/user.py [hunk 1]
```diff
@@ -36,4 +37,4 @@ class User(Representation):
 
     @property
     def name(self):
-        return self.displayname or self.gecos or self.commonname
+        return self.displayname or self.gecos or self.common_name
```
Comment (nphilipp) [inline noggin/representation/user.py]: Why this change? Is it to conform with how it's used elsewhere, e.g. in FreeIPA?

File: noggin/templates/spamcheck.html [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+              {{ _("Your account needs to be approved by an administrator.") }}
+              {{ _("You will be notified by email when the decision has been taken.") }}
+              {{ _("Thank you for your patience.") }}
+              {% elif g.current_user.status_note == "spamcheck_denied" %}
+              {{ _("Your account has been flagged as spam. If it is not correct, please contact the administrators.") }}
+              {{ _("(if it is correct, though, please stop doing that)") }}
+              {% else %}
+              {{ _("Unsupported spam status: %s, please contact the administrators", g.current_user.status_note) }}
+              {% endif %}
+            </p>
```
Comment (nphilipp) [inline noggin/templates/spamcheck.html]: Is it necessary to wrap the text in Jinja expressions/variables here? If I'm not off, we could change it to something like this to make it easier to read and maintain:

````
{% if g.current_user.status_note == "spamcheck_awaiting" %}
  <h4 class="alert-heading">Your account is being checked</h4>
  <p>
    Your account is being checked for spam likelihood, it should only take a few seconds, please wait...
  </p>
{% elif g.current_user.status_note == "spamcheck_manual" %}
  <h4 class="alert-heading">Your account requires admin activation</h4>
  <p>
    Your account needs to be approved by an administrator.
    You will be notified by email when the decision has been taken.
    Thank you for your patience.
  </p>
{% elif g.current_user.status_note == "spamcheck_denied" %}
￼  <h4 class="alert-heading">Your account is blocked</h4>
  <p>
    Your account has been flagged as spam. If it is not correct, please contact the administrators.
    (If it is correct, though, please stop doing that.)
  </p>
{% else %}
  <h4 class="alert-heading">Something went wrong</h4>
  <p>
    Unsupported spam status: {{ g.current_user.status_note }}, please contact the administrators.
  </p>
{% endif %}
````

Also, we probably should tell people how to contact administrators -- they're new to the project and won't be aware of the various channels.

File: noggin/controller/password.py [hunk 3]
```diff
@@ -138,7 +138,10 @@ def forgot_password_ask():
                 raise FormError(
                     "username", _("User %(username)s does not exist", username=username)
                 )
-            token = PasswordResetToken.from_user(user).as_string()
+            token = make_token(
+                {"sub": user.username, "lpc": user.last_password_change},
+                audience="pwdrst",
```
Comment (nphilipp) [inline noggin/controller/password.py]: Am I right assuming that the valid values of `audience` are defined in Basset? I had to think hard what `pwdrst` means (and I'm at an advantage, I see what this was before :wink:).

**Review by nphilipp — CHANGES_REQUESTED** [review]

File: noggin/representation/user.py (hunk 1 above)
Comment (abompard) [reply noggin/representation/user.py]: Hmm not really, this change isn't really necessary. I don't know what's best. I'll revert it since after all we already have `displayname` without the underscore.

File: noggin/templates/spamcheck.html (hunk 2 above)
Comment (abompard) [reply noggin/templates/spamcheck.html]: That's indeed a bit easier to read, I'll switch to that.

File: noggin/controller/password.py (hunk 3 above)
Comment (abompard) [reply noggin/controller/password.py]: No, audience values are internal, it just needs to be the same on token creation and token parsing side, to avoid token re-use. We can use whatever we want but using too long a string has disadvantages because it makes the token longer, and it ends up in the URL so there's a size limit. I could use an enum if you think it makes it clearer.

**Comment by abompard** [comment]
Thanks for the review Nils, I've updated the PR

File: noggin/controller/password.py (hunk 3 above)
Comment (nphilipp) [reply noggin/controller/password.py]: Maybe it's just my brain taking the wrong turn when coming across anything `pwd...`, maybe we could use `passreset` (which is only one character more)?

Comment (abompard) [reply noggin/controller/password.py]: I switched to an enum, it should make things clearer :-)

File: noggin/templates/spamcheck.html [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+    <div class="container section">
+      <div class="row">
+        <div class="col-md-7">
+          {{ front_page_blurb() }}
+        </div>
+        <div class="col">
+          <div class="alert alert-{% if g.current_user.status_note == 'spamcheck_denied' %}warning{% else %}info{% endif %}" role="alert">
+            {% if g.current_user.status_note == "spamcheck_awaiting" %}
+              <h4 class="alert-heading">
+                {{ _("Your account is being checked") }}
```
Comment (nphilipp) [inline noggin/templates/spamcheck.html:16]: Again, why does it have to be wrapped in a Jinja expression like this? I'm used to messages being extracted for translation from templates and replaced with localized versions when rendered, just as they are.

**Review by nphilipp — CHANGES_REQUESTED** [review]

File: noggin/templates/spamcheck.html (hunk 4 above)
Comment (nphilipp) [reply noggin/templates/spamcheck.html:16]: To clarify my question, why is this...:

````
{{ _("Your account is being checked") }}
````

...not this?

````
Your account is being checked
````

Comment (abompard) [reply noggin/templates/spamcheck.html:16]: Not sure what you mean, do you have an example? I'm following [this doc](https://jinja.palletsprojects.com/en/2.11.x/extensions/#new-style-gettext).

Comment (pypingou) [reply noggin/templates/spamcheck.html:16]: @nphilipp for translations :)

Comment (nphilipp) [reply noggin/templates/spamcheck.html:16]: I've created a PR covering "we should do l10n in templates differently" here: https://github.com/fedora-infra/noggin/issues/317

**Review by nphilipp — APPROVED** [review]
Looks good to me!
